amr-wind/equation_systems/icns/source_terms/DragForcing.cpp

```diff
@@ -0,0 +1,153 @@
+#include "amr-wind/equation_systems/icns/source_terms/DragForcing.H"
+#include "amr-wind/utilities/IOManager.H"
+
+#include "AMReX_Gpu.H"
+#include "AMReX_Random.H"
+#include "amr-wind/wind_energy/ABL.H"
+
+namespace amr_wind::pde::icns {
+
+DragForcing::DragForcing(const CFDSim& sim)
+    : m_sim(sim)
+    , m_mesh(sim.mesh())
+    , m_velocity(sim.repo().get_field("velocity"))
+{
+    amrex::ParmParse pp("DragForcing");
+    pp.query("dragCoefficient", m_drag);
+    pp.query("spongeStrength", m_spongeStrength);
+    pp.query("spongeDensity", m_spongeDensity);
+    pp.query("spongeDistanceX", m_spongeDistanceX);
+    pp.query("spongeDistanceY", m_spongeDistanceY);
+    const auto& phy_mgr = m_sim.physics_manager();
+    if (phy_mgr.contains("ABL")) {
+        const auto& abl = m_sim.physics_manager().get<amr_wind::ABL>();
+        const VelPlaneAveraging& fa_velocity =
+            abl.abl_statistics().vel_profile_coarse();
+        gpu_vel_ht.resize(fa_velocity.line_centroids().size());
+        gpu_vel_vals.resize(fa_velocity.line_average().size());
+        amrex::Gpu::copy(
+            amrex::Gpu::hostToDevice, fa_velocity.line_centroids().begin(),
+            fa_velocity.line_centroids().end(), gpu_vel_ht.begin());
+        amrex::Gpu::copy(
+            amrex::Gpu::hostToDevice, fa_velocity.line_average().begin(),
+            fa_velocity.line_average().end(), gpu_vel_vals.begin());
+    } else {
+        m_spongeStrength = 0.0;
+    }
+}
+
+DragForcing::~DragForcing() = default;
+
+void DragForcing::operator()(
+    const int lev,
+    const amrex::MFIter& mfi,
+    const amrex::Box& bx,
+    const FieldState fstate,
+    const amrex::Array4<amrex::Real>& src_term) const
+{
+    const auto& vel =
+        m_velocity.state(field_impl::dof_state(fstate))(lev).const_array(mfi);
+    bool is_terrain = this->m_sim.repo().field_exists("terrainBlank");
+    if (!is_terrain) {
+        amrex::Abort("Need terrain blanking variable to use this source term");
+    }
+    auto* const m_terrainBlank = &this->m_sim.repo().get_field("terrainBlank");
```

Comment:
more inconsistent case. Change to snake case. I will stop these comments and just recommend you check this for this PR ;)